digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/TrackSectionRoadway.scala

```diff
@@ -0,0 +1,359 @@
+                           leftLinks: Seq[ProjectLink], firstLeft: Seq[ProjectLink], restLeft: Seq[ProjectLink])
+  : ((Seq[ProjectLink], Seq[ProjectLink]), (Seq[ProjectLink], Seq[ProjectLink])) = {
+    if (rightLinks.isEmpty || leftLinks.isEmpty || (firstRight ++ firstLeft).exists(_.status == LinkStatus.NotHandled)) {
+      ((firstRight, restRight), (firstLeft, restLeft))
+    } else if (firstRight.map(_.roadwayNumber).distinct.size == firstLeft.map(_.roadwayNumber).distinct.size) {
+      val newRoadwayNumber1 = Sequences.nextRoadwayNumber
+      val newRoadwayNumber2 = if (rightLinks.head.track == Track.Combined || leftLinks.head.track == Track.Combined) newRoadwayNumber1 else Sequences.nextRoadwayNumber
+      val handledRight = continuousRoadwaySection(firstRight, newRoadwayNumber1, firstLeft)
+      val handledLeft = continuousRoadwaySection(firstLeft, newRoadwayNumber2, firstRight)
+      ((handledRight._1, handledRight._2 ++ restRight), (handledLeft._1, handledLeft._2 ++ restLeft))
+    } else {
+      val (adjustedRight, adjustedLeft) = adjustTwoTrackRoadwayNumbers(firstRight, firstLeft)
+      ((adjustedRight, restRight), (adjustedLeft, restLeft))
+    }
+  }
+
+  private def continuousRoadwaySection(seq: Seq[ProjectLink], givenRoadwayNumber: Long, opposite: Seq[ProjectLink] = Seq.empty[ProjectLink]): (Seq[ProjectLink], Seq[ProjectLink]) = {
+    val track = seq.headOption.map(_.track).getOrElse(Track.Unknown)
+    val roadType = seq.headOption.map(_.roadType.value).getOrElse(RoadType.Empty.value)
+    val status = seq.headOption.map(_.status.value).getOrElse(LinkStatus.NotHandled.value)
+
+    val (continuousProjectLinks, restProjectLinks) = {
+      val splitAddrMValue = opposite.find(_.isSplit).map(_.endAddrMValue)
+      val startAddrMValue = seq.headOption.map(_.startAddrMValue)
+      val existSplitOnOppositeTrack = opposite.exists(_.isSplit) &&
+        (splitAddrMValue.isDefined && startAddrMValue.isDefined && splitAddrMValue.get > startAddrMValue.get)
+      val continuousSection = seq.takeWhile { pl =>
+        pl.track == track && pl.roadType.value == roadType && pl.status.value == status && !pl.isSplit && !(existSplitOnOppositeTrack && splitAddrMValue.contains(pl.endAddrMValue))
+      }.sortBy(_.startAddrMValue)
+
+      if ((seq.exists(_.isSplit) || existSplitOnOppositeTrack) && seq.drop(continuousSection.size).nonEmpty)
+        (continuousSection :+ seq.drop(continuousSection.size).head, seq.drop(continuousSection.size + 1))
+      else
+        (continuousSection, seq.drop(continuousSection.size))
+    }
+
+    val assignedContinuousSection = assignRoadwayNumbersInContinuousSection(continuousProjectLinks, givenRoadwayNumber)
+    (assignedContinuousSection, restProjectLinks)
+  }
+
+  private def assignRoadwayNumbersInContinuousSection(links: Seq[ProjectLink], givenRoadwayNumber: Long): Seq[ProjectLink] = {
```

Comment:
Do we assume here that links are always in right order?

Reply:
Yes, for every process regarding addresses and roadway numbers, project links must be ordered by _startAddrMValue_.